### PR TITLE
Fix reader for large index files.

### PR DESCRIPTION
**Pull request by Gouthamve**
Currently the offsets are cast into uint32 even though the index can
grow larger than 4GiB.

Test is currently OOMing, will fix.

---
This change is [Reviewable](https://reviewable.io/reviews/prometheus/tsdb/273)

### CODE REVIEW

**Comment by Gouthamve** [comment]
Removed test as it OOMs. @fabxc take a look?

**Comment by tzz** [comment]
Is there a workaround for users currently affected by this bug? Removing or splitting files maybe?

**Comment by fabxc** [comment]
👍

**Comment by SuperQ** [comment]
LGTM

File: index/index.go [hunk 1]
```diff
@@ -143,7 +143,11 @@ func NewWriter(fn string) (*Writer, error) {
 	if err != nil {
 		return nil, err
 	}
-	defer df.Close() // close for flatform windows
+	defer df.Close() // Close for platform windows.
+
+	if err := os.RemoveAll(fn); err != nil {
+		return nil, errors.Wrap(err, "remove any existing index at path")
+	}
```
Comment (fabxc) [inline index/index.go]: Mh, do we need this? Just opening a new file in create/write mode on it should just overwrite it as expected.

**Comment by deejay1** [comment]
This doesn't seem to make it into 2.2.0-rc.0, is there a chance for it to make it for final?

File: index/index.go (hunk 1 above)
Comment (Gouthamve) [reply index/index.go]: Yep. So if there happened to be an index that is larger than the one being written it'll cause issues. Like one of my tests, I wrote a bigger one and then a smaller one, but couldn't figure out what's wrong.

Very rare edge case, but I guess it doesn't hurt to add a check.

**Comment by fabxc** [comment]
👍

**Comment by deejay1** [comment]
@fabxc Will this make it into 2.2.0 as it didn't make the cut to rc1?

**Comment by fabxc** [comment]
Mh, what makes you think this didn't make it into rc.1? Is the underlying issue not resolved.

It _should_ be in: 
* https://github.com/prometheus/prometheus/blob/v2.2.0-rc.1/vendor/vendor.json#L805
* https://github.com/prometheus/tsdb/commit/494acd307058387ced7646f9996b0f7372eaa558

**Comment by deejay1** [comment]
@fabxc Ah, stupid me, I was looking at the release/2.2 branch not the tag :/

**Comment by tzz** [comment]
After trying Prometheus 2.2.0-rc.1 which has this fix, I can confirm that large index files are handled correctly so far (1 full day of testing). I have multiple 9-15GB index files and no errors. Thank you for making this fix available.
